feat(hooks-store): new design & refactoring

shoom3301 opened the pull request:
# Summary

Design: https://www.figma.com/design/1ctikx9sFa9Ff925GRsjvC/Hook-Store

Changes:
1. Separated pure/container components
2. Simplified `HookDappContext`: now it is passed as a component prop instead of react context
3. Added localStorage persisting for hooks state
4. Added hooks editing
5. And of course, applied the new design from [Michel's PR](https://github.com/cowprotocol/cowswap/pull/4852) 

| -- | -- | -- | -- | -- |
|--------|--------|--------|--------|--------|
| <img width="493" alt="image" src="https://github.com/user-attachments/assets/62b6de85-95f6-4876-8da2-8727e4bf1c9b"> | <img width="493" alt="image" src="https://github.com/user-attachments/assets/3ed57a35-0530-4e21-900c-7fafc97057ee"> | <img width="490" alt="image" src="https://github.com/user-attachments/assets/1357cd9d-c31e-47f4-a8b6-a51100e20d55"> | <img width="489" alt="image" src="https://github.com/user-attachments/assets/f7f14fe1-81d5-4c15-a55e-a3909813fd04"> | <img width="493" alt="image" src="https://github.com/user-attachments/assets/4d10c143-c587-4f8a-a127-b1d813010247"> | 

# To Test

Please, keep in mind that the feature is not ready yet and still under feature-flag.

1. UI should corrsepond to the Figma layouts. (reordering feature is not implemented yet)
2. Added hooks should be saved in localStorage (should be applied to the form after page refresh)
3. "Build your own" hook should be editable

alfetopito commented on apps/cowswap-frontend/src/modules/hooksStore/state/hookDetailsAtom.ts:
```diff
@@ -1,10 +1,18 @@
-  postHooks: [],
-})
+export const hooksAtom = atomWithStorage<HooksStoreState>(
+  'hooks-store-atom:v1',
```
camelCase

```suggestion
  'hooksStoreAtom:v0',
```

alfetopito reviewed (commented):
Minor comments did not test it.

Need to make sure permit is tested with and without hooks to not break it again.

elena-zh commented:
Hey @shoom3301 , great job!
However, I have some issues and questions: 
1. Hooks will not be used for wraps/unwraps, so I'd delete them from the form in these cases: 
![image](https://github.com/user-attachments/assets/d1a6ca52-f3b3-440d-809b-95a88e317efb)
2. If a hook is specified in one network, it should be removed from the form once a user changes a network (or connects to a wallet in another network)
3. I've added hooks on the Hooks tab, and then decided to do a regular swap: hooks were added to the appdata (IMO, hooks should be added to the appdata only when an order is placed from the Hooks tab)
4. I think, it would be great not to allow to add the same hook twice into the pre- or post-hook section
![image](https://github.com/user-attachments/assets/5c6deeaf-691b-4bea-a877-34d72c547e97)
5. Same case as for the one above: it would be great to add a validation when a user fills in the same custom hook (with the same data in all fields)
6. If 2 different custom hooks were added, it would be great to distinguish them somehow. Maybe add a hooks fields preview on a hook hover? 
![image](https://github.com/user-attachments/assets/83f530e0-756b-4667-9858-29faad2a9d73)
7. A specified hook was not added to the appdata for a ETH-flow order
8. A specified hook was not added to the appdata in Safe
9. If I add several hooks, and run simulation for each of them, there will be only one (the latest) link with a simulation report. Is it possible to keep them all on the form? 
![щтдн щту](https://github.com/user-attachments/assets/f0f31cb2-7bc6-4a5a-aa2a-deddf7e11d01)
10. (weird) in this PR I was not able to connect to Impersonator wallet using WC
![wwc](https://github.com/user-attachments/assets/99a7d457-2ffb-43fd-8e20-a24082d70441)
11. Hook validation: IMO, it would be great to disable a possibility of adding a hook when there is nothing to claim: 
![image](https://github.com/user-attachments/assets/f0ae703d-3dfc-40e9-be3e-47ff432c22f9)
12. UI things (for @fairlighteth ): 
- there should not be 'details' and version on a hook 
![UI](https://github.com/user-attachments/assets/edf79b60-1f51-4b6b-8ed6-b9c628fb4ab6)
- there should be search for a  hook option
- it would be great to display nicer 'connect to a wallet' message on a hook details
![connect wallet first](https://github.com/user-attachments/assets/fdf5ab4e-42e8-44ef-affb-3dc941c5ecf6)
- it would be great to improve the UI on a reduced screen 
![image](https://github.com/user-attachments/assets/937c1209-bee0-4c57-a804-8c4d07244b5b)


Thanks!

shoom3301 commented:
@elena-zh thank you!

1. Fixed
2. Fixed, now the hooks state is bound to chainId + account
3. Fixed
4. Out of this task scope
5. Out of this task scope
6. As a fast solution I added title which displays a hook uid
<img width="498" alt="image" src="https://github.com/user-attachments/assets/67563aeb-d54c-4ef7-944a-509e55a47511">
7. It is intended
8. It is intended
9. Out of this task scope
10. Couldn't reproduce
11. Out of this task scope
12. I don't think so
13. Out of this task scope

shoom3301 commented:
@elena-zh "Out of this task scope" - means I don't think we should spend time on this right now, I would reiterate this after we get some feedback from users

elena-zh commented:
Hey @shoom3301 , thank you for explanation!

Some comments: 
1. If a hook should not be applied to ETH order, IMO, it would be hide it from the form once ETH (XDAI) token is selected as a sell one. 
2. Safe or other SC wallets: this should be more global change: if a hook store should not work with SC wallets, it would be nice to hide this tab when connected to it (not sure!). However, need to discuss it with the team as [here](https://cowservices.slack.com/archives/C0361CDG8GP/p1721235010655299?thread_ts=1721212169.951999&cid=C0361CDG8GP) it was said that hooks may work with SC wallets... 
3. Hook may not be  saved for an account: 
 - connect to a wallet
 - specify a hook
 - change a wallet
 - add another hook for another account
 - change a wallet to the 1st address --> the hook for that address is not saved. 

Thanks!

shoom3301 commented:
@elena-zh it's a bit complicated for now, let's keep it as it is (hooks are always disabled for eth-flow)

elena-zh commented:
> @elena-zh it's a bit complicated for now, let's keep it as it is (hooks are always disabled for eth-flow)

added this case into #4662

elena-zh reviewed (commented):
Hey @shoom3301, there are several UI and low priority issues, but I will open a separate task to address them. 
But I'd like to mention one issue that I'm concerned about: 
1. Place an order with a permittable token (not approved) through Hooks form
2. While this order is Open, go to the swap page and place an order with the same token pair.

**AR**: there are 2 possible outcomes: 
1. The app requires me to sign a permit request once again
2. The app does not allow me to place the order due to 'insufficient allowance' error
4. in both cases   I see error fetching permit  data on console. 

![permit](https://github.com/user-attachments/assets/cd807748-59fd-4bab-a512-8ecf3ee832f9)
![куйгшкуы](https://github.com/user-attachments/assets/b1f28e6f-284c-434a-ab67-1c54413d4ec3)

Could you please double check this case and LMK whether this should be addressed in the current PR or in the another task? 
Thanks

shoom3301 commented:
@elena-zh thanks! Please put it in a new task, we will take care of it later

elena-zh commented:
> @elena-zh thanks! Please put it in a new task, we will take care of it later

Now reported here: #4727 

All the rest issues can be found in #4662
